### PR TITLE
51 authentication login

### PR DESCRIPTION
**Pull request by ahmedirheem**
relates #51

### CODE REVIEW

File: server/services/auth.ts [hunk 1]
```diff
@@ -35,4 +33,37 @@ const signupService = async (userData: UserData) => {
...
+const loginService = async (userData: userLoginAttrs) => {
+  const { password, username } = userData;
+
+  await loginSchema.validateAsync(userData);
+
+  const user = await User.findOne({
+    where: { username },
+  });
+  if (!user) {
+    throw new CustomError(404, 'User Not Found');
```
Comment (alaaahmade) [inline server/services/auth.ts:45]: we can return a response like ```هناك خطأ في اسم المستخدم او كلمة المرور```
we don't want to specify the error

File: server/services/auth.ts [hunk 2]
```diff
@@ -35,4 +33,37 @@ const signupService = async (userData: UserData) => {
...
+  }
+
+  const result = await bcrypt.compare(password, user.dataValues.password);
+
+  if (!result) {
+    throw new CustomError(401, 'Invalid Email or Password');
```
Comment (alaaahmade) [inline server/services/auth.ts:51]: the website is in Arabic: 'Invalid Email or Password'

File: server/validations/loginSchema.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import Joi from 'joi';
+
+const loginSchema = Joi.object({
```
Comment (alaaahmade) [inline server/validations/loginSchema.ts:3]: the website is in Arabic way the validation messages are in English

**Review by alaaahmade — CHANGES_REQUESTED** [review]

File: server/utils/jwt/generateToken.ts [hunk 4]
```diff
@@ -1,9 +1,9 @@
-import { sign, Secret } from 'jsonwebtoken';
-import { SECRET_KEY } from '../../config';
+import { sign } from 'jsonwebtoken';
+import environment from '../../config/environment';
```
Comment (alaaahmade) [inline server/utils/jwt/generateToken.ts:2]: way import all environment
you want SECRET_KEY

**Review by alaaahmade — CHANGES_REQUESTED** [review]

File: server/interfaces/auth.ts [hunk 5]
```diff
@@ -6,3 +6,8 @@ export interface UserData {
   phone: number;
   role: string;
 }
+
+export interface userLoginAttrs {
```
Comment (Emanalabsi) [inline server/interfaces/auth.ts:10]: why not userLoginAttributes?

**Review by Emanalabsi — CHANGES_REQUESTED** [review]

File: server/interfaces/auth.ts (hunk 5 above)
Comment (ahmedirheem) [reply server/interfaces/auth.ts:10]: Because of this 
Screenshot 2023-07-01 162735

**Review by Emanalabsi — APPROVED** [review]

File: server/services/auth.ts [hunk 6]
```diff
@@ -35,4 +32,37 @@ const signupService = async (userData: UserData) => {
...
+  const result = await bcrypt.compare(password, user.dataValues.password);
+
+  if (!result) {
+    throw new CustomError(401, 'خطأ في البريد الإلكتروني أو كلمة المرور');
+  }
+
+  const userName = user.username;
+  const { id, email, phone, role, createdAt, updatedAt } = user.dataValues;
+
+  const loggedUser = {
```
Comment (alaaahmade) [inline server/services/auth.ts:56]: you can add the loggeeduser on the cookie
and add the username 

```
  const loggedUser = { id,
   username
    email,
    phone,
    role,
  };
const token = generateToken(loogedUser)```

File: server/services/auth.ts [hunk 7]
```diff
@@ -1,18 +1,15 @@
...
 import { User } from '../models';
 import { CustomError } from '../utils';
-import { signupSchema } from '../validations/schema';
 import { generateToken } from '../utils/jwt/generateToken';
 import { UserData } from '../interfaces/auth';
+import { signupSchema, loginSchema } from '../validations';
+import { userLoginAttrs } from '../interfaces/auth';
 
 const signupService = async (userData: UserData) => {
   const { username, email, password, phone, role } = userData;
```
Comment (alaaahmade) [inline server/services/auth.ts:10]: you do not want anything from this destructuring else password and username

File: server/validations/authSchema.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+const loginSchema = Joi.object({
+  username: Joi.string().alphanum().min(4).max(50).required().messages({
+    'string.empty': 'هذا الحقل لا يمكن أن يكون فارغ',
+    'string.min': 'يجب ألا يقل هذا الحقل عن 5 أحرف',
+    'string.max': 'يجب ألا يزيد هذا الحقل عن 25 حرفًا',
+    'any.required': 'اسم المسخدم حقل مطلوب',
+  }),
+  password: Joi.string().alphanum().min(4).max(50).required().messages({
+    'string.empty': 'هذا الحقل لا يمكن أن يكون فارغ',
+    'string.base': 'This field must be a password',
```
Comment (alaaahmade) [inline server/validations/authSchema.ts:24]: one in Arabic and one in English :1st_place_medal:

File: server/validations/authSchema.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
+import Joi from 'joi';
+
+const signupSchema = Joi.object({
```
Comment (alaaahmade) [inline server/validations/authSchema.ts:3]: add custom messages 
in Arabic

File: server/services/auth.ts [hunk 10]
```diff
@@ -1,18 +1,15 @@
...
 
-  const validatedData = await signupSchema.validateAsync(userData);
-
-  if (!validatedData) {
-    throw new CustomError(422, 'UnproccableContent');
-  }
+  await signupSchema.validateAsync(userData);
 
   const userExists = await User.findOne({ where: { username } });
```
Comment (alaaahmade) [inline server/services/auth.ts:14]: if the user does not Exist 
this is not server failed
it's a bad request
401

**Review by alaaahmade — CHANGES_REQUESTED** [review]
where is the unit test ?

File: server/middleware/index.ts [hunk 11]
```diff
@@ -1 +1,4 @@
-export {};
+import checkAuth from './checkAuth';
+import isLogged from './isLogged';
```
Comment (Saeed99Madi) [inline server/middleware/index.ts:2]: isLogged not necessary

**Review by Saeed99Madi — APPROVED** [review]
Great Work All

**Review by alaaahmade — APPROVED** [review]

**Pull request merged**